File: examples/kotlin/src/main/kotlin/adr/blocks/artifact/Contract.kt
// ── blocks/artifact/contract — the work product's transport (F6) ───────────
// The artifact used to be built by PERFORMED EFFECTS — which replay stubs — so 2.2's
// "the folded, replayable result of the session" was false, and a reducer change
// that corrupted artifact content while leaving State byte-identical passed every
// check on offer.
//
// Here the artifact IS State: one line per fold arm. Delivery is ONE irreversible
// effect at seal time, gated by G6 exactly as 14.3 says session-end is.

package adr.contract

import adr.blocks.artifact.ArtifactLine
import adr.spine.pure.CommandId
import adr.spine.pure.Signature
import adr.spine.pure.Timestamp
import adr.spine.pure.ToolName

sealed interface ArtifactResult : ToolResult {
    data class RecordFinding(
        override val tool: ToolName,
        val text: String,
    ) : ArtifactResult

    data class RequestSeal(override val tool: ToolName) : ArtifactResult

    data class ConfirmSeal(override val tool: ToolName) : ArtifactResult
}

sealed interface ArtifactCommand : Command {
    data class RecordFinding(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
        val text: String,
    ) : ArtifactCommand

    data class RequestSeal(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
    ) : ArtifactCommand

    data class ConfirmSeal(
        override val tool: ToolName,
        override val sig: Signature,
        override val id: CommandId,
    ) : ArtifactCommand
}

sealed interface ArtifactEffect : Effect {
    /** IRREVERSIBLE, and it fires exactly ONCE, at seal time — never once per line. */
    data class DeliverArtifact(
        override val at: Timestamp,
        val lines: List<ArtifactLine>,
    ) : ArtifactEffect
}

File: examples/typescript/src/blocks/artifact/contract.ts
// ── blocks/artifact/contract — the work product's TRANSPORT (G16) ──────────
// The artifact was built by PERFORMED EFFECTS, which replay stubs — so 2.2's
// "the folded, replayable result of the session" was false, and a reducer
// change that corrupted artifact content while leaving State byte-identical
// passed every check on offer.
//
// Here the artifact IS State. `recordFinding` folds a line. Delivery is ONE
// irreversible effect at seal time, gated by G6 exactly as 14.3 says session-end
// is — not one effect per line performed as you go.

import type { CommandBase } from "@adr/spine/pure/command";
import type { EffectBase } from "@adr/spine/pure/effect";
import type { ToolResultBase } from "@adr/spine/pure/tool-result";
import type { ArtifactLine } from "./slice";

// ── ToolResult cases ────────────────────────────────────────────────────────
export interface RecordFindingResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "recordFinding";
  readonly text: string;
}

export interface RequestSealResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "requestSeal";
}

export interface ConfirmSealResult extends ToolResultBase {
  readonly outcome: "ok";
  readonly tool: "confirmSeal";
}

export type ArtifactResult = RecordFindingResult | RequestSealResult | ConfirmSealResult;

// ── Command cases ───────────────────────────────────────────────────────────
export interface RecordFindingCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "recordFinding";
  readonly text: string;
}

export interface RequestSealCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "requestSeal";
}

export interface ConfirmSealCommand extends CommandBase {
  readonly outcome: "ok";
  readonly tool: "confirmSeal";
}

export type ArtifactCommand = RecordFindingCommand | RequestSealCommand | ConfirmSealCommand;

// ── Effect cases ────────────────────────────────────────────────────────────
/** IRREVERSIBLE, and it fires EXACTLY ONCE, at seal. The lines it carries are a
 *  value copied out of committed State, so what was delivered is what replay
 *  re-derives. */
export interface DeliverArtifact extends EffectBase {
  readonly kind: "DeliverArtifact";
  /** IRREVERSIBLE: the work product leaves the system. Its verb `confirmSeal` is
   *  the registry's other `Irreversible` row — the same fact stated at the other
   *  end of the same step (docs/DECISIONS.md:85). */
  readonly effectClass: "Irreversible";
  readonly lines: readonly ArtifactLine[];
}

export type ArtifactEffect = DeliverArtifact;

export function isArtifactResult(r: ToolResultBase): r is ArtifactResult {
  return (
    r.outcome === "ok" &&
    (r.tool === "recordFinding" || r.tool === "requestSeal" || r.tool === "confirmSeal")
  );
}
